### PR TITLE
BUG: join placeholders when to use nested snippets

### PR DESCRIPTION
**Pull request by kitagry**
Hi @hrsh7th . Thanks for the nice plugin.

I used the following snippet nested together. But `++` was deleted as the image shows.
I found that `s:Snippet.normalize` join `$1${3:++}` to `$1`.

```json
{
  "for statement": {
    "prefix": "for",
    "body": "for ${1:index} := 0; $1 < ${2:count}; $1${3:++} {\n\t$0\n}",
    "description": "Snippet for a for loop"                                                                                                                   
  }
}
```

kitad_ tmux a 2020_07_16 23_24_42

### CODE REVIEW

**Comment by hrsh7th** [comment]
Thank you!

I'm surprised to your work.
You fix the method that difficult to understand even me...

Icthink this fix is correct.
Could you add yrst case?

**Comment by kitagry** [comment]
@hrsh7th 
Thanks for the review. I added a test case.

**Comment by hrsh7th** [comment]
Sorry, I was missing the breaking change.
I'm considering that's effect.

**Comment by hrsh7th** [comment]
I considered it.
Probably we should remove placeholder normalizing in `s:Snippet.normalize`.

I will add a few comments.

File: autoload/vsnip/session/snippet.vim [hunk 1]
```diff
@@ -352,6 +352,8 @@ function! s:Snippet.normalize() abort
       endif
     endif
 
+    let self.placeholder = v:null
+    let self.text = v:null
     let self[a:node.type] = {
```
Comment (hrsh7th) [inline autoload/vsnip/session/snippet.vim:357]: Currently, this supports `placeholder` or `text` node but I decided to remove `placeholder normalizing` so please simplify it.

After this improvements, `s:Snippet.normalize` does only merge `adjacent text node`.

File: autoload/vsnip/session/snippet.vimspec [hunk 2]
```diff
@@ -239,6 +239,13 @@ Describe vsnip#snippet
       call s:expect(l:text).to_equal(l:snippet.text())
     End
 
+    It Should normalize 3
+      let l:snippet = s:Snippet.new(s:start_position, '${1:i}${2:++}')
+      let l:children = deepcopy(l:snippet.children)
+      call l:snippet.normalize()  " Don't change anything
+      call s:expect(l:snippet.children).to_equal(l:children)
+    End
```
Comment (hrsh7th) [inline autoload/vsnip/session/snippet.vimspec:247]: Good job! 👍

File: autoload/vsnip/session/snippet.vim (hunk 1 above)
Comment (kitagry) [reply autoload/vsnip/session/snippet.vim]: Thank you. I agree with your opnion.

**Comment by kitagry** [comment]
Thank you for the review!
I simplified `s:Snippet.normalize`.

**Comment by hrsh7th** [comment]
LGTM.
Is this can be merge?

**Comment by kitagry** [comment]
OK!
Thanks for the review!

**Comment by hrsh7th** [comment]
Thank you!

**Pull request merged**